### PR TITLE
 VCF masking

### PR DESCRIPTION
**Pull request by jeromekelleher**
After discussing things with @szhan and the bottlenecks he's experiencing, this is a proposal we came up with for how we can provide a flexible interface for masking out sites and samples in the VCF output. Usage:
```python
ts = msprime.sim_ancestry(5, random_seed=1, sequence_length=100)
ts = msprime.sim_mutations(ts, rate=0.1, random_seed=1234)
site_mask = np.zeros(ts.num_sites, dtype=int)
site_mask[::4] = 1
sample_mask = np.zeros(ts.num_samples, dtype=int)
sample_mask[0] = 1
sample_mask[1] = 1

def sample_mask(variant):
    sample_mask = np.zeros(ts.num_samples, dtype=int)
    sample_mask[variant.site.id % ts.num_samples] = 1
    return sample_mask

ts.write_vcf(sys.stdout, site_mask=site_mask, sample_mask=sample_mask)
```

Here, in this contrived example, we're masking out every 4th site, and have a sample mask that depends on the site. This is done by having ``sample_mask`` optionally be a callable, which takes the variant as input and returns the relevant mask.

Output:
```
##fileformat=VCFv4.2
##source=tskit 0.4.1
##FILTER=<ID=PASS,Description="All filters passed">
##contig=<ID=1,length=100>
##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
#CHROM  POS     ID      REF     ALT     QUAL    FILTER  INFO    FORMAT  tsk_0   tsk_1   tsk_2   tsk_3   tsk_4
1       0       0       A       C       .       PASS    .       GT      .|.     .|.     .|.     .|.     .|.
1       1       1       A       T,C     .       PASS    .       GT      0|.     0|0     0|0     0|1     0|0
1       2       2       T       G       .       PASS    .       GT      1|0     .|1     1|1     1|0     1|1
1       4       3       G       A,T     .       PASS    .       GT      2|0     0|.     0|1     1|0     2|1
1       7       4       T       C       .       PASS    .       GT      .|.     .|.     .|.     .|.     .|.
1       11      5       A       T       .       PASS    .       GT      0|1     1|0     0|.     0|0     0|0
1       13      6       A       T       .       PASS    .       GT      1|0     1|1     1|1     .|0     1|1
1       15      7       A       C       .       PASS    .       GT      0|1     0|0     0|0     0|.     0|0
1       16      8       G       A       .       PASS    .       GT      .|.     .|.     .|.     .|.     .|.
1       17      9       A       T       .       PASS    .       GT      1|0     1|1     1|1     1|0     1|.
1       26      10      T       G,A     .       PASS    .       GT      .|0     2|1     1|2     2|0     1|2
1       27      11      A       C       .       PASS    .       GT      0|.     0|0     0|0     0|0     1|0
1       28      12      G       C       .       PASS    .       GT      .|.     .|.     .|.     .|.     .|.
1       37      13      C       T       .       PASS    .       GT      0|0     0|.     0|0     0|1     0|0
1       39      14      T       G       .       PASS    .       GT      0|0     1|0     .|0     0|0     0|0
1       41      15      G       A       .       PASS    .       GT      1|0     1|1     1|.     1|0     1|1
1       45      16      G       C       .       PASS    .       GT      .|.     .|.     .|.     .|.     .|.
1       46      17      G       A       .       PASS    .       GT      0|0     1|0     0|0     0|.     0|0
1       47      18      G       T       .       PASS    .       GT      0|0     0|0     1|0     0|0     .|0
1       50      19      A       T       .       PASS    .       GT      0|0     0|0     0|1     0|0     0|.
1       53      20      T       G       .       PASS    .       GT      .|.     .|.     .|.     .|.     .|.
1       54      21      T       C       .       PASS    .       GT      0|.     0|0     0|1     1|0     0|1
1       56      22      T       A       .       PASS    .       GT      1|0     .|1     1|1     1|0     1|1
1       58      23      G       C       .       PASS    .       GT      0|1     0|.     0|0     0|1     0|0
1       61      24      C       T,A     .       PASS    .       GT      .|.     .|.     .|.     .|.     .|.
1       62      25      C       G       .       PASS    .       GT      0|0     0|0     1|.     0|0     0|0
1       64      26      A       G       .       PASS    .       GT      1|0     0|1     0|0     .|0     1|0
1       68      27      G       A       .       PASS    .       GT      1|0     0|1     0|0     0|.     1|0
1       69      28      G       A       .       PASS    .       GT      .|.     .|.     .|.     .|.     .|.
1       71      29      C       G       .       PASS    .       GT      0|0     0|0     0|0     0|1     0|.
1       73      30      G       A       .       PASS    .       GT      .|1     0|0     0|0     0|1     0|0
1       77      31      T       C       .       PASS    .       GT      0|.     0|0     0|0     0|0     1|0
1       78      32      C       G,A,T   .       PASS    .       GT      .|.     .|.     .|.     .|.     .|.
1       79      33      A       T       .       PASS    .       GT      0|0     1|.     0|0     0|0     0|0
1       80      34      A       C       .       PASS    .       GT      1|0     0|1     .|0     0|0     0|0
1       88      35      C       A       .       PASS    .       GT      0|0     0|0     0|.     0|0     0|1
1       91      36      G       A       .       PASS    .       GT      .|.     .|.     .|.     .|.     .|.
1       92      37      C       T       .       PASS    .       GT      1|0     0|1     0|0     0|.     1|0
1       93      38      G       T       .       PASS    .       GT      0|1     0|0     0|0     0|1     .|0
1       94      39      G       C       .       PASS    .       GT      0|1     0|0     0|0     0|1     0|.
1       95      40      G       A       .       PASS    .       GT      .|.     .|.     .|.     .|.     .|.
1       96      41      G       A       .       PASS    .       GT      0|.     0|0     0|1     1|0     0|1
1       97      42      C       T       .       PASS    .       GT      1|1     .|1     0|0     0|1     1|0
1       99      43      A       C,G,T   .       PASS    .       GT      3|1     0|.     2|0     0|1     3|0
```

We haven't tested this for perf at all yet, so it could be a significant regression and need a little refactoring to make sure the common case is still fast.

Thoughts?

### CODE REVIEW

**Comment by jeromekelleher** [comment]
Oh yes, this should also give us support for missing data in the VCF output, as a nice side-effect.

**Comment by jeromekelleher** [comment]
For performance evaluation we would:

1. Get a large tree sequence (say 100k samples, a few thousand sites - something that takes 10s of seconds to output the VCF for), and time vcf output on main.
2. Run this code with no masks, and time.
3. Run this code with a constant sample mask (say masking out even samples or something) and time
4. Run this code with a site-dependent sample mask (like the one above) and time.

**Comment by petrelharp** [comment]
Sounds good - my only suggestion is that shouldn't masks be Boolean?

**Comment by jeromekelleher** [comment]
> Sounds good - my only suggestion is that shouldn't masks be Boolean?

Yeah, agreed.

**Comment by szhan** [comment]
I did some preliminary performance evaluation.

```
import timeit

my_setup = '''
import sys

import tskit
import msprime

import numpy as np

ts = msprime.sim_ancestry(1e6, sequence_length=1e5, random_seed=1234)
ts = msprime.sim_mutations(ts, rate=1e-4, random_seed=5678)

site_mask = np.zeros(ts.num_sites, dtype=int)
site_mask[::4] = 1

sample_mask = np.zeros(ts.num_samples, dtype=int)
sample_mask[::4] = 1

def get_site_specific_sample_mask(variant):
    sample_mask = np.zeros(ts.num_samples, dtype=int)
    sample_mask[variant.site.id % ts.num_samples] = 1
    return sample_mask
'''

print("Test 1: No mask")
print(
    timeit.timeit(
        stmt='''ts.write_vcf(open('/dev/null', 'w'), site_mask=None, sample_mask=None)''',
        setup=my_setup,
        number=10
    )
)

print("Test 2: Sample mask")
print(
    timeit.timeit(
        stmt='''ts.write_vcf(open('/dev/null', 'w'), site_mask=None, sample_mask=sample_mask)''',
        setup=my_setup,
        number=10
    )
)

print("Test 3: Site mask")
print(
    timeit.timeit(
        stmt='''ts.write_vcf(open('/dev/null', 'w'), site_mask=site_mask, sample_mask=None)''',
        setup=my_setup,
        number=10
    )
)

print("Test 4: Site-dependent sample mask")
print(
    timeit.timeit(
        stmt='''ts.write_vcf(open('/dev/null', 'w'), site_mask=None, sample_mask=get_site_specific_sample_mask)''',
        setup=my_setup,
        number=10
    )
)
```

Test 1 (no mask): 656.5268213189993
Test 2 (sample mask): 852.0801203749998 (30% slower)
Test 3 (site mask): 716.2780362409994 (9% slower)
Test 4 (sample and site mask): 810.0072484789998 (23% slower)

**Comment by jeromekelleher** [comment]
Very interesting, thanks @szhan. Can you tell me what the timing for the current code in ``main`` is, for reference?

**Comment by szhan** [comment]
I'm seeing 490.993569683 when doing this ``timeit`` run.

```
print("Test 0: No mask (original)")
print(
    timeit.timeit(
        stmt='''ts.write_vcf(open('/dev/null', 'w'))''',
        setup=my_setup,
        number=10
    )
)
```

**Comment by szhan** [comment]
Is the dip in speed acceptable?

**Comment by jeromekelleher** [comment]
> Is the dip in speed acceptable?

Not for the nominal "no mask" case, no. I can update easily enough to make a "fast-path" for no masking, which should hopefully have exactly the same performance as the old code. Will do when I get a chance.

**Comment by jeromekelleher** [comment]
TODO:

1. Error checking on mask inputs
2. Test that pysam is OK with our missing data encoding
3. Documentation

**Comment by jeromekelleher** [comment]
This should be pretty much ready to go now, can you take a look please @benjeffery?

File: python/tskit/trees.py [hunk 1]
```diff
@@ -5463,6 +5477,23 @@ def write_vcf(
 
             $ tskit vcf example.trees | bcftools view -O b > example.bcf
 
+
+        The ``sample_mask`` argument provides a general way to mask out
+        parts of the output, which can be helpful when simulating missing
+        data. In this (contrived) example, we create a sample mask function
+        that marks one genotype missing in each variant in a regular
+        patter:
```
Comment (benjeffery) [inline python/tskit/trees.py:5485]:
```suggestion
        pattern:
```

File: python/tskit/vcf.py [hunk 2]
```diff
@@ -204,7 +223,23 @@ def write(self, output):
                 end="\t",
                 file=output,
             )
-            gt_array[indexes] = variant.genotypes + ord("0")
+            # NOTE: when we suppose missing data we should be able to
```
Comment (benjeffery) [inline python/tskit/vcf.py:226]:
```suggestion
            # NOTE: when we support missing data we should be able to
```

File: python/tskit/vcf.py [hunk 3]
```diff
@@ -204,7 +223,23 @@ def write(self, output):
...
+            # simply add ``and not variant.has_missing_data`` here.
+            # Probably OK to take the perf hit in making the missing
+            # data case to with the more general sample masking case.
```
Comment (benjeffery) [inline python/tskit/vcf.py:229]: ???
```suggestion
            # data case deal with the more general sample masking case.
```

**Review by benjeffery — APPROVED** [review]
LGTM Well tested, just a couple of typos.

**Comment by jeromekelleher** [comment]
Awesome, thanks. Marked for merging.

**Pull request merged**